### PR TITLE
refactor: update table metadata in single txn

### LINKED ISSUE #1715: Update table metadata values in a batch atomically
### What type of enhancement is this?

Refactor

### What does the enhancement do?

When doing region failover, the updating of metadata values is guarded by a distributed lock. It does not robust enough. For example, the lock lease could be expired in the middle of one's updating, letting others to start updating concurrently. For now, we set the lease of the distributed lock to 10 seconds, which is long enough here to get the job done. Maybe we should introduce "version" companion values to these two metadata values, and use ETCD transaction request to update them?

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

update table metadata in single txn

### Benchmark

```
➜  greptimedb git:(refactor/table-metadata-manager-part2) ✗ ./target/debug/greptime cli bench --etcd-addr=127.0.0.1:2379 --count=1000  
2023-08-17T03:43:46.470364Z  INFO greptime: short_version: 0.3.2, full_version: greptimedb-refactor/table-metadata-manager-part2-0745c89
2023-08-17T03:43:46.470453Z  INFO greptime: command line arguments
2023-08-17T03:43:46.470468Z  INFO greptime: argument: ./target/debug/greptime
2023-08-17T03:43:46.470481Z  INFO greptime: argument: cli
2023-08-17T03:43:46.470491Z  INFO greptime: argument: bench
2023-08-17T03:43:46.470501Z  INFO greptime: argument: --etcd-addr=127.0.0.1:2379
2023-08-17T03:43:46.470511Z  INFO greptime: argument: --count=1000
2023-08-17T03:43:48.555916Z  INFO cmd::cli::bench: TableMetadataBencher: creating 1000 table metadata, average operation cost: 2.05 ms
2023-08-17T03:43:49.157239Z  INFO cmd::cli::bench: TableMetadataBencher: getting 1000 table info and region routes, average operation cost: 0.60 ms
2023-08-17T03:43:52.548820Z  INFO cmd::cli::bench: TableMetadataBencher: renaming 1000 table, average operation cost: 0.82 ms
2023-08-17T03:43:55.464835Z  INFO cmd::cli::bench: TableMetadataBencher: deleting 1000 table metadata, average operation cost: 0.81 ms
```

```
➜  greptimedb git:(refactor/table-metadata-manager-part2) ✗ ./target/debug/greptime cli bench --etcd-addr=127.0.0.1:2379 --count=2000
2023-08-17T03:44:51.661023Z  INFO greptime: short_version: 0.3.2, full_version: greptimedb-refactor/table-metadata-manager-part2-0745c89
2023-08-17T03:44:51.661110Z  INFO greptime: command line arguments
2023-08-17T03:44:51.661125Z  INFO greptime: argument: ./target/debug/greptime
2023-08-17T03:44:51.661138Z  INFO greptime: argument: cli
2023-08-17T03:44:51.661148Z  INFO greptime: argument: bench
2023-08-17T03:44:51.661158Z  INFO greptime: argument: --etcd-addr=127.0.0.1:2379
2023-08-17T03:44:51.661168Z  INFO greptime: argument: --count=2000
2023-08-17T03:44:55.832656Z  INFO cmd::cli::bench: TableMetadataBencher: creating 2000 table metadata, average operation cost: 2.05 ms
2023-08-17T03:44:57.069445Z  INFO cmd::cli::bench: TableMetadataBencher: getting 2000 table info and region routes, average operation cost: 0.62 ms
2023-08-17T03:45:03.670236Z  INFO cmd::cli::bench: TableMetadataBencher: renaming 2000 table, average operation cost: 0.83 ms
2023-08-17T03:45:09.478045Z  INFO cmd::cli::bench: TableMetadataBencher: deleting 2000 table metadata, average operation cost: 0.81 ms
```

```
➜  greptimedb git:(refactor/table-metadata-manager-part2) ✗ ./target/debug/greptime cli bench --etcd-addr=127.0.0.1:2379 --count=5000
2023-08-17T03:45:32.682439Z  INFO greptime: short_version: 0.3.2, full_version: greptimedb-refactor/table-metadata-manager-part2-0745c89
2023-08-17T03:45:32.682529Z  INFO greptime: command line arguments
2023-08-17T03:45:32.682545Z  INFO greptime: argument: ./target/debug/greptime
2023-08-17T03:45:32.682557Z  INFO greptime: argument: cli
2023-08-17T03:45:32.682568Z  INFO greptime: argument: bench
2023-08-17T03:45:32.682578Z  INFO greptime: argument: --etcd-addr=127.0.0.1:2379
2023-08-17T03:45:32.682588Z  INFO greptime: argument: --count=5000
2023-08-17T03:45:43.122005Z  INFO cmd::cli::bench: TableMetadataBencher: creating 5000 table metadata, average operation cost: 2.05 ms
2023-08-17T03:45:46.081609Z  INFO cmd::cli::bench: TableMetadataBencher: getting 5000 table info and region routes, average operation cost: 0.59 ms
2023-08-17T03:46:00.793505Z  INFO cmd::cli::bench: TableMetadataBencher: renaming 5000 table, average operation cost: 0.80 ms
2023-08-17T03:46:14.457716Z  INFO cmd::cli::bench: TableMetadataBencher: deleting 5000 table metadata, average operation cost: 0.81 ms
```

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
close #1715

### CODE REVIEW

File: src/common/meta/src/rpc/router.rs [hunk 1]
```diff
@@ -98,6 +102,48 @@ pub struct TableRoute {
...
+            let _ = peers.insert(p.clone());
+        });
+    region_routes
+        .iter()
+        .flat_map(|x| x.follower_peers.iter())
+        .for_each(|p| {
+            let _ = peers.insert(p.clone());
+        });
+    let mut peers = peers.into_iter().collect::<Vec<_>>();
+    peers.sort_by_key(|x| x.id);
```
Comment (fengjiachun) [inline src/common/meta/src/rpc/router.rs:142]: Why do we need sorting?

File: src/common/meta/src/rpc/router.rs [hunk 2]
```diff
@@ -98,6 +102,48 @@ pub struct TableRoute {
...
+        .for_each(|p| {
+            let _ = peers.insert(p.clone());
+        });
+    region_routes
+        .iter()
+        .flat_map(|x| x.follower_peers.iter())
+        .for_each(|p| {
+            let _ = peers.insert(p.clone());
+        });
+    let mut peers = peers.into_iter().collect::<Vec<_>>();
```
Comment (fengjiachun) [inline src/common/meta/src/rpc/router.rs:141]: Better to iter `region_routes` only once, the code is as follows:

```
let mut peers = region_routes
        .iter()
        .flat_map(|x| x.leader_peer.iter().chain(x.follower_peers.iter()))
        .collect::<HashSet<_>>()
        .into_iter()
        .cloned()
        .collect::<Vec<_>>();
```

File: src/common/meta/src/rpc/router.rs [hunk 3]
```diff
@@ -251,6 +297,18 @@ impl TryFrom<PbTableRouteValue> for TableRoute {
     }
 }
 
+impl TryFrom<TableRoute> for PbTableRouteValue {
+    type Error = error::Error;
```
Comment (fengjiachun) [inline src/common/meta/src/rpc/router.rs:301]: Add an empty line

File: src/meta-srv/src/procedure/alter_table.rs [hunk 4]
```diff
@@ -75,11 +73,57 @@ impl AlterTableProcedure {
...
+            ))
+            .await
+            .context(TableMetadataManagerSnafu)?;
+
+        ensure!(
+            exist,
+            error::TableNotFoundSnafu {
+                name: request.table_ref().to_string()
+            }
+        );
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/alter_table.rs:116]: We can execute a batch exists (or a txn) to reduce a call to the remote storage.

File: src/common/meta/src/rpc/router.rs (hunk 1 above)
Comment (WenyXu) [reply src/common/meta/src/rpc/router.rs:142]: Reduces unnecessary updates.  e.g., `[3, 1] != [1, 3]` 

We update the `DatanodeTableKey` like the following: 
```rust 
        for (datanode, regions) in new_region_distribution.into_iter() {
            if let Some(current_region) = current_region_distribution.get(&datanode) {
                // Updates if need.
                if *current_region != regions {
                    let key = DatanodeTableKey::new(datanode, table_id);
                    let raw_key = key.as_raw_key();
                    let val = DatanodeTableValue::new(table_id, regions).try_as_raw_value()?;
                    opts.push(TxnOp::Put(raw_key, val));
                }
            } else {
                // New datanodes
                let key = DatanodeTableKey::new(datanode, table_id);
                let raw_key = key.as_raw_key();
                let val = DatanodeTableValue::new(table_id, regions).try_as_raw_value()?;
                opts.push(TxnOp::Put(raw_key, val));
            }
        }
```

**Comment by WenyXu** [comment]
@fengjiachun @MichaelScofield PTAL

File: src/meta-srv/src/procedure/alter_table.rs [hunk 5]
```diff
@@ -79,41 +79,65 @@ impl AlterTableProcedure {
...
+        let exists = manager
+            .table_name_manager()
+            .batch_exists(compare_keys)
             .await
-            .context(TableMetadataManagerSnafu)?;
+            .context(error::TableMetadataManagerSnafu)?;
+
+        for handler in handlers {
+            handler(&exists)?;
+        }
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/alter_table.rs:139]: Apologies for not thoroughly reviewing the code yesterday. I misunderstood the requirement and believed it was to ensure everything exists. This change is overly complex. Let's revert back to the original code. ^_^

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/key/table_region.rs [hunk 6]
```diff
@@ -15,19 +15,21 @@
...
-use crate::kv_backend::KvBackendRef;
-use crate::rpc::store::{CompareAndPutRequest, MoveValueRequest};
-use crate::DatanodeId;
+use crate::error::{InvalidTableMetadataSnafu, Result, SerdeJsonSnafu};
+use crate::key::TableMetaKey;
+use crate::{impl_table_meta_key, impl_table_meta_value, DatanodeId};
 
 pub type RegionDistribution = BTreeMap<DatanodeId, Vec<RegionNumber>>;
 
+#[deprecated(
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key/table_region.rs:29]: upgrade program still need it

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**